File: backend/app/schemas/kb.py
from __future__ import annotations

"""Pydantic schemas for Knowledge Base (Lesson 4)."""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field
from typing import Literal


class KBDocument(BaseModel):
    id: Optional[str] = None
    text: str = Field(min_length=1)
    metadata: Optional[Dict[str, Any]] = None


class KBIngestRequest(BaseModel):
    collection: str = Field(default="kb_main", min_length=1)
    chunk: bool = True
    max_chars: int = 600
    overlap: int = 80
    chunk_strategy: Literal["window", "punctuation"] = "window"
    delimiters: str | None = None
    documents: List[KBDocument]


class KBIngestResponse(BaseModel):
    collection: str
    inserted_ids: List[str]
    chunks_added: int


class KBQueryRequest(BaseModel):
    collection: str = Field(default="kb_main", min_length=1)
    query: str = Field(min_length=1)
    n_results: int = 5


class KBMatch(BaseModel):
    id: str | None = None
    text: str
    metadata: Optional[Dict[str, Any]] = None
    distance: float | None = None


class KBQueryResponse(BaseModel):
    collection: str
    query: str
    matches: List[KBMatch]


class KBDeleteRequest(BaseModel):
    collection: str = Field(default="kb_main", min_length=1)
    ids: List[str]


class KBDeleteResponse(BaseModel):
    collection: str
    deleted: int


class KBItem(BaseModel):
    id: str
    text: str | None = None
    metadata: Dict[str, Any] | None = None


class KBListResponse(BaseModel):
    collection: str
    total: int
    items: List[KBItem]

File: frontend/src/api/kb.ts
/**
 * API service for Knowledge Base (Lesson 4)
 */
import { apiClient } from './client'

export interface KBDocument {
  id?: string
  text: string
  metadata?: Record<string, any>
}

export interface KBIngestRequest {
  collection?: string
  chunk?: boolean
  max_chars?: number
  overlap?: number
  documents: KBDocument[]
}

export interface KBIngestResponse {
  collection: string
  inserted_ids: string[]
  chunks_added: number
}

export interface KBQueryRequest {
  collection?: string
  query: string
  n_results?: number
}

export interface KBMatch {
  id?: string
  text: string
  metadata?: Record<string, any>
  distance?: number
}

export interface KBQueryResponse {
  collection: string
  query: string
  matches: KBMatch[]
}

export interface KBDeleteRequest {
  collection?: string
  ids: string[]
}

export interface KBDeleteResponse {
  collection: string
  deleted: number
}

export const kbApi = {
  async ingest(data: KBIngestRequest): Promise<KBIngestResponse> {
    const response = await apiClient.post<KBIngestResponse>('/kb/ingest', data)
    return response.data
  },

  async search(data: KBQueryRequest): Promise<KBQueryResponse> {
    const response = await apiClient.post<KBQueryResponse>('/kb/search', data)
    return response.data
  },

  async delete(data: KBDeleteRequest): Promise<KBDeleteResponse> {
    const response = await apiClient.post<KBDeleteResponse>('/kb/delete', data)
    return response.data
  },

  async list(params: { collection: string; limit?: number; offset?: number }): Promise<{
    collection: string
    total: number
    items: { id: string; text?: string; metadata?: Record<string, any> }[]
  }> {
    const response = await apiClient.get('/kb/items', { params })
    return response.data
  }
}
